torch/ao/sparsity/sparsifier.py

```diff
@@ -0,0 +1,167 @@
+
+import abc
+import copy
+
+import torch
+from torch import nn
+from torch.nn.utils import parametrize
+
+from .parametrization import MulBy
+
+SUPPORTED_MODULES = {
+    nn.Linear
+}
+
+def _module_to_path(model, layer, prefix=''):
+    for name, child in model.named_children():
+        new_name = prefix + '.' + name
+        if child is layer:
+            return new_name
+        child_path = _module_to_path(child, layer, prefix=new_name)
+        if child_path is not None:
+            return child_path
+    return None
+
+def _path_to_module(model, path):
+    path = path.split('.')
+    for name in path:
+        model = getattr(model, name, None)
+        if model is None:
+            return None
+    return model
+
+
+class BaseSparsifier(abc.ABC):
+    r"""Base class for all sparsifiers.
+
+    Abstract methods that need to be implemented:
+
+    - step: Function to compute a new mask for all keys in the `module_groups`.
+
+    Args:
+        - model [nn.Module]: model to configure. The model itself is not saved
+            but used for the state_dict saving / loading.
+        - config [list]: configuration elements could either be instances of
+            nn.Module or dict maps. The dicts must have a key 'module' with the
+            value being an instance of a nn.Module.
+        - defaults [dict]: default configurations will be attached to the
+            configuration. Only the keys that don't exist in the `config` will
+            be updated.
+
+    Example::
+
+        >>> config = [model.layer1, {'module': model.linear2, 'sparsity_level': 0.5}]
+        >>> defaults = {'sparsity_level': 0.7}
+        >>> # model.layer1 will have `sparsity_level` = 0.7 (getting default)
+        >>> sparsifier = BaseSparsifier(config, defaults)
+    """
+    def __init__(self, model, config, defaults):
+        super().__init__()
+        self.config = config
+        self.defaults = defaults
+        if self.defaults is None:
+            self.defaults = dict()
+
+        self.state = {}
+        self.module_groups = []
+        self.enable_mask_update = False
+
+        self.model = model
+        # If no config -- try getting all the supported layers
+        if self.config is None:
+            # Add all models to the config
+            self.config = []
+            stack = [model]
```

Comment:
Why do we need a while loop here? Cant we just check for this with a for loop followed by an if?.

Also, in this case, we should have the user specify the list of modules, instead of passing the entire model